[docs,qemu] Add QEMU OT documentation (#28553)
lowRISC/opentitan: third_party/qemu/README.md

# QEMU

For instructions on setting up QEMU for local development and troubleshooting steps, see the [setup guide](./setup.md).

## Introduction

> [QEMU](https://www.qemu.org/) is an open-source project which primarily aims to support virtualization, user-mode emulation and full-system (machine) emulation flows.

Reviewer: ```suggestion
[QEMU](https://www.qemu.org/) is an open-source emulator project for functional full-system or user-mode emulation for a variety of machines and architectures.
```
I think it is important to introduce from the outset that QEMU is a generic emulator project that attempts to do functional emulation which may not be the most accurate.

OpenTitan is an example of full-system emulation, where we emulate an OpenTitan RISC-V 32-bit machine (equivalent to a "top") to support running OpenTitan programs on any generic supported host machine, such as an x86-64 PC.

> QEMU's predominant advantage is its speed: internally, QEMU uses a dynamic translation backend called [TCG](https://www.qemu.org/docs/master/devel/tcg.html) (_Tiny Code Generator_) which optimizes translation of blocks of guest (e.g. RISC-V) instructions to equivalent host (e.g. x86-64) instructions to execute.

Reviewer: The phrasing seems a bit off - the dynamic translator doesn't optimise the translation of blocks, it translates blocks of guest code into optimised native code.

Also a minor nit: I think "The predominant/primary advantage of QEMU is its speed" reads better.

OpenTitan's emulated machines are currently developed on a [fork](https://github.com/lowRISC/qemu) of upstream QEMU to better facilitate the implementation of non-standard Ibex and OpenTitan features.
There are currently two main emulation targets: Earlgrey 1.0.0, and Darjeeling (targeting the moving `master` branch).
You can refer to the QEMU repository's documentation for more information on the development and usage of the QEMU OpenTitan machines - this page instead focuses on providing details for testing software with QEMU.

## Scope

A more comprehensive description of supported features and devices can be found in the relevant QEMU [Earlgrey](https://github.com/lowRISC/qemu/blob/ot-9.2.0/docs/opentitan/earlgrey.md) and [Darjeeling](https://github.com/lowRISC/qemu/blob/ot-9.2.0/docs/opentitan/darjeeling.md) documentation pages.
Note that these may not be up-to-date compared to the implementation, and as such the source files should be treated as the root source of truth.
Furthermore, QEMU releases are pinned in Bazel, meaning that the supported features may differ compared to when this release was taken (see the [setup guide](./setup.md)).

> Currently, only the Earlgrey machine (`ot-earlgrey`) has any support within OpenTitanLib for test integration, primarily on the `earlgrey_1.0.0` branch. As of the time of writing, a **non-exhaustive** list of some of the major features not emulated by QEMU include:

Reviewer: I think there is a list of unimplemented and missing features in our QEMU docs already. Maybe we should not duplicate that list here to avoid this possibly getting out-of-sync, and point people there instead. Though I think the general notes about timing-sensitive or synchronous protocols and the like could remain here to illustrate deficiencies when compared to other testing/verification methods, up to you though.

Reviewer: I think the important information that should be made upfront for someone taking a quick glance would be that some devices are not implemented at all (can give a short list of some key ones), a lot may have missing functionality and so it is advised to read the QEMU documentation for that device (where other invariants and technicalities might be documented too), and that some communication protocols are not modelled entirely right (e.g. I2C).

If you could compress the information down into general points like these, it would give a good overview of the project's state that would not require much updating besides after overhauls, but it is not the biggest issue if we have some documentation duplicated across repos.

When using a hardware timer (e.g. the `aon_timer` or `rv_timer`), the timer will expire after _approximately_ the correct amount of time, but will not be exactly accurate.

QEMU has mechanisms for pacing the virtual CPU to match hardware timers, but in general it should be expected that both hardware timers and instruction counts are not correct in QEMU.
This can be a particular challenge for security countermeasures and software mechanisms that opt to set tight timing thresholds.
One potential solution is modifying the `icount` shift parameter, which informs the TCG's instruction counting feature for system emulation.
When providing an icount shift of `N`, the virtual CPU will execute one instruction every `2^N` nanoseconds of virtual time, meaning that changing this parameter may allow emulation to "pass" tighter timing checks at the cost of less accurate timing overall.
By default, an `icount` of 6 is used to align execution time with wall-clock time.

### The TLB and ePMP

Hardware will often use a Translation Lookaside Buffer (TLB) as a high-speed cache to accelerate translation from virtual addresses to physical addresses.

> QEMU mimics this concept with its own address translation cache that it calls the [TLB](https://www.qemu.org/docs/master/devel/tcg.html#mmu-emulation), which speeds up translation from guest virtual addresses to either virtual addresses of host RAM or MMIO operations for emulated devices.

Reviewer: Guest Virtual to Host Virtual is true, but the equivalent which I think makes more sense here is Guest Physical.
```suggestion
QEMU mimics this concept with its own address translation cache that it calls the [TLB](https://www.qemu.org/docs/master/devel/tcg.html#mmu-emulation), which it uses to speed up translation of guest virtual addresses to guest RAM or to MMIO regions of emulated devices.
```

Reviewer: AFAIK Your understanding is correct, but I think the data and code structure of the QEMU TLB is an implementation detail that I do not think needs much elaborating, just that it performs the functions of an MMU in translating Guest Virtual to Guest Phyiscal addresses, which could be in plain memory or memory-mapped regions that dispatch to I/O.
